File: 260. Single Number III/singleNumber.py
from typing import List

class Solution:
  def singleNumber(self, nums: List[int]) -> List[int]:
    xor = 0
    for  n in nums:
      xor ^= n

    diffBit = 1
    while not xor & diffBit:
      diffBit = diffBit << 1

    a, b = 0, 0
    for n in nums:
      if diffBit & n:
        a = a ^ n
      else:
        b = b ^ n

    return [a, b]
  

nums = [1,2,1,3,2,5]
solution = Solution()
print('singleNumber: ', solution.singleNumber(nums))

File: 260. Single Number III/singleNumber.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
const singleNumber = (nums) => {
  let xor = 0;
  for (let i = 0; i < nums.length; i++) {
    xor ^= nums[i];
  }

  let diffBit = 1;
  while (!(diffBit & xor)) {
    diffBit = diffBit << 1;
  }

  let a = 0;
  let b = 0;
  for (let i = 0; i < nums.length; i++) {
    if (diffBit & nums[i]) {
      a = a ^ nums[i];
    } else {
      b = b ^ nums[i];
    }
  }

  return [a, b];
};


const nums = [1, 2, 1, 3, 2, 5];
console.log('singleNumber: ', singleNumber(nums))
